File: python/110-Balanced-Binary-Tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    # Top down recursion approach -- requires child subtree heights be redundantly calculated
    # Can be improved by using bottom-up recursion
    # Space O(n)  May check all nodes if tree is skewed
    # Time O(nlogn)
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if root == None:
            return True
        elif abs(self.getHeight(root.left) - self.getHeight(root.right)) <= 1:
            return self.isBalanced(root.left) and self.isBalanced(root.right)
        else:
            return False

    def getHeight(self, root) -> int:
        if root == None:
            return 0
        else:
            return 1 + max(self.getHeight(root.left), self.getHeight(root.right))

File: c#/110-Balanced-Binary-Tree.cs
#region DFS
// Time O(n) - n is the number of nodes in the tree
// Space O(h) - h is the height of the tree
//  Best Case (Balanced Tree): O(log(n))
//  Worst Case (Skewed Tree): O(n)
public class Solution {
    public bool IsBalanced(TreeNode root) {
        return GetDepth(root) != -1;
    }

    public int GetDepth(TreeNode root) {
        if (root == null) return 0;

        int leftD = GetDepth(root.left);
        if (leftD == -1) {
            return -1;
        }

        int rightD = GetDepth(root.right);
        if (rightD == -1) {
            return -1;
        }

        if (Math.Abs(leftD - rightD) > 1) {
            return -1;
        }

        return Math.Max(leftD, rightD) + 1;
    }
}
#endregion

// Time		O(n*log(n))
// Space	O(N)
public class Solution
{
    public bool IsBalanced(TreeNode root)
    {
        if (root == null)
        {
            return true;
        }
        else if (Math.Abs(Measure(root.left) - Measure(root.right)) < 2)
        {
            return IsBalanced(root.left) && IsBalanced(root.right);
        }
        return false;
    }

    public int Measure(TreeNode root)
    {
        if (root == null)
        {
            return 0;
        }
        return 1 + Math.Max(Measure(root.left), Measure(root.right));
    }
}

//   Definition for a binary tree node.
public class TreeNode
{
    public int val;
    public TreeNode left;
    public TreeNode right;

    public TreeNode(int val = 0, TreeNode left = null, TreeNode right = null)
    {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}
